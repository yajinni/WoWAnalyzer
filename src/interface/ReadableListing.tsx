import React, { ReactNode } from 'react';

interface Props {
  children: ReactNode | ReactNode[];
  groupType?: 'and' | 'or';
}

const ReadableListing = ({ children, groupType = 'and' }: Props) => {
  const numItems = React.Children.count(children);
  const results: React.ReactNode[] = [];
  React.Children.forEach(children, (child, index) => {
    const isFirst = index === 0;
    if (!isFirst) {
      const isLast = index === numItems - 1;
      if (isLast) {
        results.push(` ${groupType} `);
      } else {
        results.push(', ');
      }
    }
    results.push(child);
  });
  return <>{results}</>;
};

export default ReadableListing;
