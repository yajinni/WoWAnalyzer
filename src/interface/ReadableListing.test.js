import React from 'react';
import { shallow } from 'enzyme';

import ReadableListing from './ReadableListing';

const Dummy = () => <div />;

describe('ReadableList', () => {
  it('doesn\'t mutate a single item', () => {
    expect(shallow(<ReadableListing><Dummy /></ReadableListing>)).toMatchSnapshot();
  });
  it('merges two items with an "and"', () => {
    expect(shallow(<ReadableListing><Dummy /><Dummy /></ReadableListing>)).toMatchSnapshot();
  });
  it('merges three items with a "," and an "and"', () => {
    expect(shallow(<ReadableListing><Dummy /><Dummy /><Dummy /></ReadableListing>)).toMatchSnapshot();
  });
  it('merges four items with two "," and an "and"', () => {
    expect(shallow(<ReadableListing><Dummy /><Dummy /><Dummy /><Dummy /></ReadableListing>)).toMatchSnapshot();
  });
});
