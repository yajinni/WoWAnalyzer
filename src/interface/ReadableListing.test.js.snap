// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadableList doesn't mutate a single item 1`] = `
<Fragment>
  <Dummy />
</Fragment>
`;

exports[`ReadableList merges four items with two "," and an "and" 1`] = `
<Fragment>
  <Dummy />
  , 
  <Dummy />
  , 
  <Dummy />
   and 
  <Dummy />
</Fragment>
`;

exports[`ReadableList merges three items with a "," and an "and" 1`] = `
<Fragment>
  <Dummy />
  , 
  <Dummy />
   and 
  <Dummy />
</Fragment>
`;

exports[`ReadableList merges two items with an "and" 1`] = `
<Fragment>
  <Dummy />
   and 
  <Dummy />
</Fragment>
`;
